### PR TITLE
Adding Library table filtering feature

### LINKED ISSUE #119: Add option to display library information in AMDirT viewer
Currently, the AMDirT viewer command just displays the information in the respective sample tables. However, it would enhance its use very much when it could also additionally show selected columns of the library tables at the same time. This would be particularly important if one would like to select samples based on sequencing depth or library layout etc., too.

I know that this is not trivial because joining the two tables won't always be a one-to-one join but could also lead to a one-to-many join. However, I think it would be still worth it.

### PR DESCRIPTION
**Pull request by maxibor**
Adding the possibility to also filter on libraries.
Will require a documentation update for the tutorials.
Demonstration below.

Will close #119 

[amdirt_library_filtering_demo.webm](https://github.com/SPAAM-community/AMDirT/assets/8630837/1494d451-692e-4525-82de-2281da76d40e)

### CODE REVIEW

**Comment by maxibor** [comment]
Convert tool is now fixed, as well as the issue described in #127

**Comment by maxibor** [comment]
Also add download with sratoolkit to close #138

**Comment by jfy133** [comment]
Testing results:

- [ ] Library filtering
- [ ] Convert tool
- [ ]  SRAToolkit

File: AMDirT/core/__init__.py [hunk 1]
```diff
@@ -318,15 +301,16 @@ def prepare_accession_table(
         supported_archives (list): list of supported archives
     """
 
-    selected_libraries = get_libraries(
-        table_name=table_name,
-        samples=samples,
-        libraries=libraries,
-        supported_archives=supported_archives,
-    )
+    # libraries = get_libraries(
```
Comment (jfy133) [inline AMDirT/core/__init__.py:304]: Could be removed if not used anymore?

File: AMDirT/viewer/streamlit.py [hunk 2]
```diff
@@ -159,28 +184,64 @@ def parse_args():
...
+                        )
+                    except KeyError:
+                        lib_mod = pd.DataFrame(lib_sel["selected_rows"])
+
+                    if st.form_submit_button(
+                        "Validate library selection", type="primary"
+                    ):
+                        if len(lib_mod) == 0:
+                            st.error(
+                                "You didn't select any library! Please select at least one library."
```
Comment (jfy133) [inline AMDirT/viewer/streamlit.py:218]:
```suggestion
                                "You didn't select any libraries! Please select at least one library."
```

**Review by jfy133 — COMMENTED** [review]
Nothing jumps out at me after looking at the code (based on my amateur knowledge).

Onto testing now.

**Comment by jfy133** [comment]
@maxibor testing of viewer was fine, however I'm getting lots of errors with convert.

I try with both the samples and libraries files in the following zip

[Uploading amdirt-testing.zip…]()

And I get the following erros for both (but chaning the JSON schema it's validating against...)

```
$ AMDirT convert --sratoolkit samples.tsv ancientmetagenome-hostassociated
Traceback (most recent call last):
  File "/home/james/bin/miniconda3/envs/amdirt-dev/bin/AMDirT", line 33, in <module>
    sys.exit(load_entry_point('AMDirT', 'console_scripts', 'AMDirT')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/click/decorators.py", line 34, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/git/spaam-community/AMDirT/AMDirT/cli.py", line 179, in convert
    run_convert(**kwargs, **ctx.obj)
  File "/home/james/git/spaam-community/AMDirT/AMDirT/convert/__init__.py", line 71, in run_convert
    dataset_valid.append(v.validate_schema())
                         ^^^^^^^^^^^^^^^^^^^
  File "/home/james/git/spaam-community/AMDirT/AMDirT/validate/domain/__init__.py", line 235, in validate_schema
    for err in validator.iter_errors(self.dataset_json):
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/jsonschema/validators.py", line 354, in iter_errors
    for error in errors:
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/jsonschema/_legacy_validators.py", line 115, in items_draft6_draft7_draft201909
    yield from validator.descend(item, items, path=index)
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/jsonschema/validators.py", line 402, in descend
    for error in errors:
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/jsonschema/_validators.py", line 305, in properties
    yield from validator.descend(
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/jsonschema/validators.py", line 402, in descend
    for error in errors:
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/jsonschema/_validators.py", line 284, in ref
    yield from validator._validate_reference(ref=ref, instance=instance)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/jsonschema/validators.py", line 432, in _validate_reference
    resolved = self._resolver.lookup(ref)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/referencing/_core.py", line 586, in lookup
    retrieved = self._registry.get_or_retrieve(uri)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/referencing/_core.py", line 336, in get_or_retrieve
    resource = registry._retrieve(uri)
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/jsonschema/validators.py", line 120, in _warn_for_remote_retrieve
    return referencing.Resource.from_contents(json.load(response))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/james/bin/miniconda3/envs/amdirt-dev/lib/python3.11/site-packages/referencing/_core.py", line 129, in from_contents
    raise exceptions.CannotDetermineSpecification(contents)
referencing.exceptions.CannotDetermineSpecification: {'enum': ['Afghanistan', 'Albania', 'Algeria', 'American Samoa', 'Andorra', 'Angola', 'Anguilla', 'Antarctica', 'Antigua and Barbuda', 'Arctic Ocean', 'Argentina', 'Armenia', 'Aruba', 'Ashmore and Cartier Islands', 'Atlantic Ocean', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Baltic Sea', 'Baker Island', 'Bangladesh', 'Barbados', 'Bassas da India', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bolivia', 'Borneo', 'Bosnia and Herzegovina', 'Botswana', 'Bouvet Island', 'Brazil', 'British Virgin Islands', 'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon', 'Canada', 'Cape Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Chile', 'China', 'Christmas Island', 'Clipperton Island', 'Cocos Islands', 'Colombia', 'Comoros', 'Cook Islands', 'Coral Sea Islands', 'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Curacao', 'Cyprus', 'Czech Republic', 'Democratic Republic of the Congo', 'Denmark', 'Djibouti', 'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia', 'Eswatini', 'Ethiopia', 'Europa Island', 'Falkland Islands (Islas Malvinas)', 'Faroe Islands', 'Fiji', 'Finland', 'France', 'French Guiana', 'French Polynesia', 'French Southern and Antarctic Lands', 'Gabon', 'Gambia', 'Gaza Strip', 'Georgia', 'Germany', 'Ghana', 'Gibraltar', 'Glorioso Islands', 'Greece', 'Greenland', 'Grenada', 'Guadeloupe', 'Guam', 'Guatemala', 'Guernsey', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Heard Island and McDonald Islands', 'Honduras', 'Hong Kong', 'Howland Island', 'Hungary', 'Iceland', 'India', 'Indian Ocean', 'Indonesia', 'Iran', 'Iraq', 'Ireland', 'Isle of Man', 'Israel', 'Italy', 'Jamaica', 'Jan Mayen', 'Japan', 'Jarvis Island', 'Jersey', 'Johnston Atoll', 'Jordan', 'Juan de Nova Island', 'Kazakhstan', 'Kenya', 'Kerguelen Archipelago', 'Kingman Reef', 'Kiribati', 'Kosovo', 'Kuwait', 'Kyrgyzstan', 'Laos', 'Latvia', 'Lebanon', 'Lesotho', 'Liberia', 'Libya', 'Liechtenstein', 'Line Islands', 'Lithuania', 'Luxembourg', 'Macau', 'Madagascar', 'Malawi', 'Malaysia', 'Maldives', 'Mali', 'Malta', 'Marshall Islands', 'Martinique', 'Mauritania', 'Mauritius', 'Mayotte', 'Mediterranean Sea', 'Mexico', 'Micronesia', 'Midway Islands', 'Moldova', 'Monaco', 'Mongolia', 'Montenegro', 'Montserrat', 'Morocco', 'Mozambique', 'Myanmar', 'Namibia', 'Nauru', 'Navassa Island', 'Nepal', 'Netherlands', 'New Caledonia', 'New Zealand', 'Nicaragua', 'Niger', 'Nigeria', 'Niue', 'Norfolk Island', 'North Korea', 'North Macedonia', 'North Sea', 'Northern Mariana Islands', 'Norway', 'Oman', 'Pacific Ocean', 'Pakistan', 'Palau', 'Palmyra Atoll', 'Panama', 'Papua New Guinea', 'Paracel Islands', 'Paraguay', 'Peru', 'Philippines', 'Pitcairn Islands', 'Poland', 'Portugal', 'Puerto Rico', 'Qatar', 'Republic of the Congo', 'Reunion', 'Romania', 'Ross Sea', 'Russia', 'Rwanda', 'Saint Helena', 'Saint Kitts and Nevis', 'Saint Lucia', 'Saint Pierre and Miquelon', 'Saint Vincent and the Grenadines', 'Samoa', 'San Marino', 'Sao Tome and Principe', 'Saudi Arabia', 'Senegal', 'Serbia', 'Seychelles', 'Sierra Leone', 'Singapore', 'Sint Maarten', 'Slovakia', 'Slovenia', 'Solomon Islands', 'Somalia', 'South Africa', 'South Georgia and the South Sandwich Islands', 'South Korea', 'South Sudan', 'Southern Ocean', 'Spain', 'Spratly Islands', 'Sri Lanka', 'State of Palestine', 'Sudan', 'Suriname', 'Svalbard', 'Sweden', 'Switzerland', 'Syria', 'Taiwan', 'Tajikistan', 'Tanzania', 'Tasman Sea', 'Thailand', 'Timor-Leste', 'Togo', 'Tokelau', 'Tonga', 'Trinidad and Tobago', 'Tromelin Island', 'Tunisia', 'Turkey', 'Turkmenistan', 'Turks and Caicos Islands', 'Tuvalu', 'USA', 'Uganda', 'Ukraine', 'United Arab Emirates', 'United Kingdom', 'Uruguay', 'Uzbekistan', 'Vanuatu', 'Venezuela', 'Viet Nam', 'Virgin Islands', 'Wake Island', 'Wallis and Futuna', 'West Bank', 'Western Sahara', 'Yemen', 'Zambia', 'Zimbabwe', 'Unknown']}
```

**Review by jfy133 — CHANGES_REQUESTED** [review]
Getting errors in `convert` :( see above

**Comment by jfy133** [comment]
[amdirt-error.zip](https://github.com/SPAAM-community/AMDirT/files/13937853/amdirt-error.zip)

**Comment by jfy133** [comment]
OK, I reinstalled everything from scratch (rather than `pip install -e .` within teh cloned branch in the directory, and now it's working. Convert looks to be working well.

Last question: usage just says `SAMPLES` . But description says 'samples and libraries', which is correct? Giving a library table doesn't work  for me - gives a schemas check error.

```
Usage: AMDirT convert [OPTIONS] SAMPLES TABLE_NAME

  Converts filtered samples and libraries tables to eager, ameta, taxprofiler, and fetchNGS input tables
  
  SAMPLES: path to filtered AncientMetagenomeDir samples tsv file
  TABLE_NAME: name of table to convert

```

**Comment by jfy133** [comment]
@maxibor :point_up:

**Comment by maxibor** [comment]
@jfy133 Time to have another look ;)

File: AMDirT/cli.py [hunk 3]
```diff
@@ -110,6 +135,20 @@ def viewer(ctx, no_args_is_help=True, **kwargs):
     type=click.Path(exists=True),
     help="(Optional) JSON file listing AncientMetagenomeDir tables",
 )
+@click.option(
+    "--libraries",
+    type=click.Path(readable=True, file_okay=True, dir_okay=False, exists=True),
+    help=("(Optional) Path to libraries table"),
```
Comment (jfy133) [inline AMDirT/cli.py:141]:
```suggestion
    help=("(Optional) Path to a pre-filtered libraries table"),
```

File: AMDirT/convert/__init__.py [hunk 4]
```diff
@@ -36,33 +41,96 @@ def run_convert(
     """Run the AMDirT conversion application to input samplesheet tables for different pipelines
 
     Args:
-        tables (str): Path to JSON file listing tables
         samples (str): Path to AncientMetagenomeDir filtered samples tsv file
+        libraries(str): Optional path to AncientMetagenomeDir filtered libraries tsv file
```
Comment (jfy133) [inline AMDirT/convert/__init__.py:45]:
```suggestion
        libraries(str): Optional path to AncientMetagenomeDir pre-filtered libraries tsv file
```

**Review by jfy133 — APPROVED** [review]
`convert` working for me now!

In addition to the typo/left-over comment in my original review, I would suggest augmenting the main help description of `convert` to the following:


```
$ AMDirT convert --help
Usage: AMDirT convert [OPTIONS] SAMPLES TABLE_NAME

  Converts pre-filtered samples and libraries tables to eager, ameta, taxprofiler, and fetchNGS input tables.

Note: When supplying a pre-filtered libraries table to `--libraries`, the corresponding sample table is required!

  SAMPLES: path to filtered AncientMetagenomeDir samples tsv file
  TABLE_NAME: name of table to convert
  ```
  
  But otherwise LGTM after those three changes :)

**Pull request merged**